cardano-node/src/Cardano/CLI/Delegation/PoolRegistry.hs

```diff
@@ -0,0 +1,301 @@
+
+-- | Bech32-encoded address.
+data PledgeAddress = PledgeAddress
+  { paHumanReadable :: Bech32.HumanReadablePart
+  , paDataPart :: Bech32.DataPart
+  }
+
+-- | Given a stake pool registry root and a submission file,
+--   perform full validation of relative file structure
+--   & JSON content of the submission.
+validateRegistrySubmission
+  :: RegistryRoot
+  -> SubmissionFile
+  -> ExceptT CliError IO Submission
+validateRegistrySubmission (RegistryRoot root) (SubmissionFile fp) = do
+  -- Directory structure/access checks.
+  --
+  afp <- lift $ Dir.makeAbsolute fp
+  unlessM (lift $ Dir.doesFileExist afp) $
+    checks ["Submission doesn't exist: " <> pack afp]
+  unlessM (lift $ Dir.readable <$> Dir.getPermissions afp) $
+    checks ["Submission not readable: " <> pack afp]
+  unlessM (lift $ Dir.doesDirectoryExist root) $
+    checks ["Registry root doesn't exist: " <> pack afp]
+
+  let (dir', file) = Path.splitFileName afp
+      dir = Path.dropTrailingPathSeparator dir'
+      (parentParentDir, parentDir) = Path.splitFileName dir
+
+  checks $
+    ["Submission not in 'registry' directory"
+    | parentDir /= "registry"] <>
+    ["Submission not properly relative the registry root"
+    | Path.dropTrailingPathSeparator root
+      /= Path.dropTrailingPathSeparator parentParentDir]
+
+  let (id, ext) = Path.splitExtension file
+      sigPath = dir </> id <.> "sig"
+
+  unlessM (lift $ Dir.doesFileExist sigPath) $
+    checks ["Missing signature file: " <> pack sigPath]
+  unlessM (lift $ Dir.readable <$> Dir.getPermissions sigPath) $
+    checks ["Signature file not readable: " <> pack sigPath]
+  checks $
+    ["Submission filename doesn't have .json extension: " <> pack file
+    | ext /= ".json" ]
+
+  pubkey <- case validatePublicKey (pack id) of
+    Left e -> err $
+      ["Submission filename (actually public key) invalid: " <> e]
+    Right x -> pure x
+
+  contents <- lift $ BSL.fromStrict <$> BS.readFile afp
+  case AE.eitherDecode contents of
+    Left e -> err $ ["Submission content:\n" <> pack e]
+    Right entry -> do
+
+      checks $
+        [ pack $ "Internal 'id' doesn't match filename: "
+          <> show (unId $ sId entry)  <> " vs. " <> show (unId pubkey)
+        | sId entry /= pubkey ]
+
+      -- All cheap checks done, let's do the expensive one now.
+      dupTickers <- withExceptT (PoolRegistryValidationErrors fp . (:[]) .
+                               ("While reading all registry entries: "<>)) $
+        duplicateTickers dir
+
+      checks $
+        [ "Duplicate tickers: " <>
+          pack (unpack $ Text.intercalate ", " $ unTicker <$> dupTickers)
+        | not (null dupTickers) ]
+
+      -- All valid.
+      pure entry
+
+ where
+   checks :: [Text] -> ExceptT CliError IO ()
+   checks [] = ExceptT . pure . Right $ ()
+   checks es = err es
+
+   err :: [Text] -> ExceptT CliError IO a
+   err es = ExceptT . pure . Left  $ PoolRegistryValidationErrors fp es
```

Comment:
Wouldn't the use of [throwE](https://hackage.haskell.org/package/transformers-0.5.6.2/docs/Control-Monad-Trans-Except.html#v:throwE) makes the above code a bit simpler / more idiomatic?

Reply:
Good point, will do!

Reply:
You _may_ also be able to make use of `firstExceptT` in the above code.